core/pkg/server/handler.go

```diff
@@ -1090,13 +1078,11 @@ func (h *Handler) handlePartialHistoryAsync(request *service.PartialHistoryReque
 	}
 }
 
-// The main entry point for partial history records in the non-shared mode.
+// handlePartialHistorySync updates the run history in non-shared mode.
 //
-// This collects partial history records until a full history record is
-// received. This is the synchronous version of handlePartialHistory.
-// In this mode, the client will assign a step number to the history record.
-// The client is responsible for ensuring that the step number is monotonically
-// increasing.
+// In this mode, we are the only process writing to this run.
```

Comment:
this is not accurate we could have multiple processes writing to the same run, but we are responsible for managing the step.

We currently only have communication on the same node, hence if the user's processes stay in the localhost we can manage the step for them, but for multi node case we need the backend to manage the steps.

Maybe in the future we will get away of managing the step and have just infra manage the _step, while we treat this user provide step just as another metric. Probably will help with all the resume /rewind requests, as users could overwrite earlier steps

Reply:
but there is only one wandb-core process, right?

Reply:
yep!

Reply:
that's what I mean by "process" in the comment :)